tsdb/head.go

```diff
@@ -373,9 +378,17 @@ func NewHead(r prometheus.Registerer, l log.Logger, wal *wal.WAL, opts *HeadOpti
 		opts.SeriesCallback = &noopSeriesLifecycleCallback{}
 	}
 
-	es, err := NewCircularExemplarStorage(opts.NumExemplars, r)
-	if err != nil {
-		return nil, err
+	// ugly hack to get the previous default size
+	if opts.MaxExemplars == 0 {
+		opts.MaxExemplars = 100000
+	}
+	// Always start with a noopExemplarStorage now since ApplyConfig is run right after TSDB starts up.
```

Comment:
If exemplars is enabled, wouldn't `opts.EnableExemplarStorage && opts.MaxExemplars > -1` always be true? And `EnableExemplarStorage` wont change without a restart, so that state wont change in `ApplyConfig`.

Reply:
> If exemplars is enabled, wouldn't opts.EnableExemplarStorage && opts.MaxExemplars > -1 always be true?

If users were exprimenting with exemplars and then wanted to stop using them for a while to evaluate their findings, reduce memory usage, etc., they could set `MaxExempalrs` to a negative # to return to using a `noopExemplarStorage` without needing to restart Prometheus.

> And EnableExemplarStorage wont change without a restart, so that state wont change in ApplyConfig.

Yes, but overall it could not have been set to true at startup but a user can still add `MaxExemplars` to the config file.

Reply:
This is in `NewHead`, which is called only on startup, and I see in `main.go` we have this

```go
if cfgFile.StorageConfig.ExemplarsConfig.MaxExemplars > 0 {
	cfg.tsdb.MaxExemplars = cfgFile.StorageConfig.ExemplarsConfig.MaxExemplars
}
```
Which means we ignore anything that is `<0`, hence this will always `true` at the moment. So this needs to change in `main.go` I think to accept any number?